Add swizzling support (xyzw and rgba)

EkardNT opened the pull request:
Adds swizzling support for up to the first 4 elements of a vector.

One thing that cgmath does that I haven't yet is put swizzling behind a feature flag. They claim that adding swizzling will add 0.6 MB to the library size, but then (correctly, AFAIK) point out that rustc will remove unused functions when building, so I'm not sure why they bothered with the flag. I didn't see a reason to, but I will add it if you want.

EkardNT committed:
Implement swizzling.

EkardNT committed:
Revert unintended whitespace changes.

EkardNT commented:
Here's the size results. It is indeed about 600KB, which is kind of neat considering I didn't look at cgmath's implementation at all :).

| Size of libaljabar.rlib | Before swizzling | After swizzling |
|-------------------------|------------------|-----------------|
| ... in debug build      | 245.1 KB         | 799.5 KB        |
| ... in release build    | 196 KB           | 750.4 KB        |

EkardNT committed:
Comment cleanup, removed remnants of my failed attempt to add doc comments to generated swizzle methods.

maplant commented:
I really, really like this. Although I'm conflicted about putting it behind a feature gate. I'm not sure that 550 KB is really worth it to selectively gate some features, but on the other hand half a meg is half a meg.

One comment: could you please add a [compile_fail example](https://doc.rust-lang.org/nightly/edition-guide/rust-2018/rustdoc/documentation-tests-can-now-compile-fail.html) to back up the claim that calling a component accessor on too small of a vec will cause a compilation failure?

maplant commented on src/lib.rs:
````diff
@@ -262,6 +262,174 @@ impl<T, const N: usize> Vector<T, {N}> {
+/// ```ignore
+/// let v = vector![1, 2, 3, 4].xxzz();
+/// ```
+pub trait Swizzle<T> {
````
I believe that `Swizzle` must require `Sized` in order to make it object-unsafe. As of right now you can create Swizzle trait objects and I believe that could cause undefined behaviour.

EkardNT replied:
Interesting, what would the UB be? Understanding trait objects and object safety is a weak part in my Rust knowledge.

Adding `: Sized` bound.

maplant replied:
I don't think it's UB. Fortunately the rust designers figured this one out correctly.

My issue is that without the Sized bound you would be able to write the following: 
```rust
fn take_swizzle<T>(swiz: &dyn Swizzle<T>) -> T
where
    T: Copy
{
    swiz.z()
}
```
attempts to pass in a vec to this function with N < 3 causes an outs of bounds error as one would expect:
```index out of bounds: the len is 1 but the index is 2```
This error is not terribly helpful, so I think for now it would be wise to disallow this for now. 
Making this sized removes the ability to turn this into a trait object.

maplant commented:
Ah, one last thing: please add a section in the README on this

EkardNT committed:
Address PR feedback.
- Add :Sized bound to Swizzle.
- Add compile_fail tests.
- Add swizzle support description to Readme.

EkardNT commented:
Latest commit:
- Add :Sized bound to Swizzle.
- Add compile_fail tests.
- Add swizzle support description to Readme.

maplant commented:
What's the rational for requiring `T: Copy` instead of `Clone`?

maplant commented:
@EkardNT so I've been experimenting with this and I realized that there's no real reason to put these methods as a separate trait - you can just put them into an `impl Vector` and achieve what I think is a much preferable result. I'll adjust this code and get it merged tonight
